### PR TITLE
New plugin: Munition Controller

### PR DESCRIPTION
**Pull request by Aingar**
Extension of #203 put into its own plugin.
Also allows specifying shipclasses which shouldn't be tracked by a given projectile.
Main usage: Disabling tracking of Gunboat missiles against snubs.

### CODE REVIEW

File: Plugins/Public/munitioncntl/Main.cpp [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+
+void LoadSettings()
+{
+	returncode = DEFAULT_RETURNCODE;
+
+	INI_Reader ini;
+
+	char szCurDir[MAX_PATH];
+	GetCurrentDirectory(sizeof(szCurDir), szCurDir);
+	string scPluginCfgFile = string(szCurDir) + "\\flhook_plugins\\munitioncntl.cfg";
```
Comment (Lazrius) [inline Plugins/Public/munitioncntl/Main.cpp:72]: Use raw string literal

File: Plugins/Public/munitioncntl/Main.cpp [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+///////////////////////////////////////////////////////////////////////////////////////////////////////////////
+//Functions
+///////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+void ProcessGuided(FLPACKET_CREATEGUIDED& createGuidedPacket)
+{
+	uint ownerType;
+	pub::SpaceObj::GetType(createGuidedPacket.iOwner, ownerType);
+	if (!(ownerType & (OBJ_FIGHTER | OBJ_FREIGHTER | OBJ_TRANSPORT | OBJ_GUNBOAT | OBJ_CRUISER | OBJ_CAPITAL))) //GetTarget throws an exception for non-ship entities.
+		return;
```
Comment (Lazrius) [inline Plugins/Public/munitioncntl/Main.cpp:134]: Add braces

File: Plugins/Public/munitioncntl/Main.cpp [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+///////////////////////////////////////////////////////////////////////////////////////////////////////////////
+//Functions
+///////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+void ProcessGuided(FLPACKET_CREATEGUIDED& createGuidedPacket)
+{
+	uint ownerType;
+	pub::SpaceObj::GetType(createGuidedPacket.iOwner, ownerType);
+	if (!(ownerType & (OBJ_FIGHTER | OBJ_FREIGHTER | OBJ_TRANSPORT | OBJ_GUNBOAT | OBJ_CRUISER | OBJ_CAPITAL))) //GetTarget throws an exception for non-ship entities.
```
Comment (Lazrius) [inline Plugins/Public/munitioncntl/Main.cpp:133]: I suggest adding these to a global constexpr

File: Plugins/Public/munitioncntl/Main.cpp [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+		}
+		case TRACK_NOALERT:
+		{
+			createGuidedPacket.iTargetId = 0; // prevents the 'incoming missile' warning client-side
+		}
+	}
+
+}
+
+void __stdcall CreateGuided(uint& iClientID, FLPACKET_CREATEGUIDED& createGuidedPacket)
```
Comment (Lazrius) [inline Plugins/Public/munitioncntl/Main.cpp:174]: Is the stdcall needed?

File: Plugins/Public/munitioncntl/Main.cpp [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+BOOL WINAPI DllMain(HINSTANCE hinstDLL, DWORD fdwReason, LPVOID lpvReserved)
+{
+	srand((uint)time(0));
+	// If we're being loaded from the command line while FLHook is running then
+	// set_scCfgFile will not be empty so load the settings as FLHook only
+	// calls load settings on FLHook startup and .rehash.
+	if (fdwReason == DLL_PROCESS_ATTACH)
+	{
+		if (set_scCfgFile.length() > 0)
+			LoadSettings();
```
Comment (Lazrius) [inline Plugins/Public/munitioncntl/Main.cpp:43]: Add braces

File: Plugins/Public/munitioncntl/Main.cpp [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+	TRACK_ALERT,
+	TRACK_NOALERT,
+	NOTRACK_NOALERT
+};
+
+void LoadSettings();
+
+BOOL WINAPI DllMain(HINSTANCE hinstDLL, DWORD fdwReason, LPVOID lpvReserved)
+{
+	srand((uint)time(0));
```
Comment (Lazrius) [inline Plugins/Public/munitioncntl/Main.cpp:36]: We don't use rand in this dll.

File: Plugins/Public/munitioncntl/Main.cpp [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+}
+
+/// Hook will call this function after calling a plugin function to see if we the
+/// processing to continue
+EXPORT PLUGIN_RETURNCODE Get_PluginReturnCode()
+{
+	return returncode;
+}
+
+bool bPluginEnabled = true;
```
Comment (Lazrius) [inline Plugins/Public/munitioncntl/Main.cpp:58]: Unused variable

File: Plugins/Public/munitioncntl/Main.cpp [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+{
+	returncode = DEFAULT_RETURNCODE;
+
+	INI_Reader ini;
+
+	char szCurDir[MAX_PATH];
+	GetCurrentDirectory(sizeof(szCurDir), szCurDir);
+	string scPluginCfgFile = string(szCurDir) + "\\flhook_plugins\\munitioncntl.cfg";
+	if (ini.open(scPluginCfgFile.c_str(), false))
+	{
```
Comment (Lazrius) [inline Plugins/Public/munitioncntl/Main.cpp:74]: Invert the if to reduce scope and make it more readable

File: Plugins/Public/munitioncntl/Main.cpp [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+							blacklistedTrackingTypesBitmap |= OBJ_CRUISER;
+						if (typeStr.find("capital") != string::npos)
+							blacklistedTrackingTypesBitmap |= OBJ_CAPITAL;
+						if (typeStr.find("guided") != string::npos)
+							blacklistedTrackingTypesBitmap |= OBJ_GUIDED;
+						if (typeStr.find("mine") != string::npos)
+							blacklistedTrackingTypesBitmap |= OBJ_MINE;
+					}
+				}
+				mapTrackingByObjTypeBlacklistBitmap[missileArch] = blacklistedTrackingTypesBitmap;
```
Comment (Lazrius) [inline Plugins/Public/munitioncntl/Main.cpp:118]: Check that we got both values populated. If either is invalid console log and do not add.

File: Plugins/Public/munitioncntl/Main.cpp [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+	}
+	else if (setNoTrackingAlertProjectiles.count(createGuidedPacket.iMunitionId)) // for 'dumbified' seeker missiles, disable alert, used for flaks and snub dumbfires
+	{
+		tracking = TRACK_NOALERT;
+	}
+	else if (mapTrackingByObjTypeBlacklistBitmap.count(createGuidedPacket.iMunitionId)) // disable tracking for selected ship types
+	{
+		uint targetType;
+		pub::SpaceObj::GetType(createGuidedPacket.iTargetId, targetType);
+		const auto& blacklistedShipTypeTargets = mapTrackingByObjTypeBlacklistBitmap.at(createGuidedPacket.iMunitionId);
```
Comment (Lazrius) [inline Plugins/Public/munitioncntl/Main.cpp:152]: Do we validate that this is in the list?

**Review by Lazrius — CHANGES_REQUESTED** [review]

File: Plugins/Public/munitioncntl/Main.cpp (hunk 4 above)
Comment (Aingar) [reply Plugins/Public/munitioncntl/Main.cpp:174]: All other packet hooks are defined with __stdcall, so I added it to this one as well. The hook definitely works.

Are there any risks involved?

File: Plugins/Public/munitioncntl/Main.cpp (hunk 10 above)
Comment (Aingar) [reply Plugins/Public/munitioncntl/Main.cpp:152]: Within the if above it, using the .count() method.

**Review by Lazrius — APPROVED** [review]

**Review by dsyalex — APPROVED** [review]
We've discussed the force push and it seems Aingar pushed another commit to the PR and then force-pushed to remove it, leaving this in the state in which Laz approved it.

Approving per Laz's approval

**Pull request merged**